source/opt/loop_fusion_pass.cpp

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2018 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "opt/loop_fusion_pass.h"
+
+#include "opt/ir_context.h"
+#include "opt/loop_descriptor.h"
+#include "opt/loop_fusion.h"
+#include "opt/register_pressure.h"
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status LoopFusionPass::Process(ir::IRContext* c) {
+  bool modified = false;
+  ir::Module* module = c->module();
+
+  // Process each function in the module
+  for (ir::Function& f : *module) {
+    modified |= ProcessFunction(&f);
+  }
+
+  return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
+}
+
+bool LoopFusionPass::ProcessFunction(ir::Function* function) {
+  auto c = function->context();
+  ir::LoopDescriptor& ld = *c->GetLoopDescriptor(function);
+
+  for (auto& loop_0 : ld) {
+    for (auto& loop_1 : ld) {
```

Comment:
Wouldn't the "try to fuse with a loop starting at the merge block" still result in looking through all the loops in the loop descriptor to see if there is one starting at the merge block?

Reply:
Yes, it will.  The issue is not functional.  It is purely a compile time performance issue.  If you can pick out the only loop that it could possibly be fused with then we don't have to check every loop.  No need to change it now, just add a TODO.

Reply:
Alright, will do.